### ISSUE TITLE
Hangs during getAllProcesses under Linux

### ISSUE THREAD

**Issue by Gobliins** [profesorfalken/jProcesses#10]
When retrieveing all Processes during the parseList option it seems like a thread is waiting endless..
However when i interrupt the wait with "strg + c" sometimes i get this exception:

´
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 1
        at org.jutils.jprocesses.info.UnixProcessesService.loadFullCommandData(UnixProcessesService.java:160)
        at org.jutils.jprocesses.info.UnixProcessesService.parseList(UnixProcessesService.java:69)
        at org.jutils.jprocesses.info.AbstractProcessesService.getList(AbstractProcessesService.java:56)
        at org.jutils.jprocesses.info.AbstractProcessesService.getList(AbstractProcessesService.java:43)
        at org.jutils.jprocesses.JProcesses.getProcessList(JProcesses.java:41)
        at processlist.JProcessLister.listProcesses(JProcessLister.java:12)
        at fetcher.ProcessFetcher.main(ProcessFetcher.java:17)
`

Sometimes it just exits...

My env data:
OS: SUSE Linux Enterprise Server 11 (x86_64)
JDK: 1.8_031

**Comment by profesorfalken** [comment]
Hello, 

What is the output in you system when you perform:
_ps -e -o pid,command_

Can you give me an example of your output?

Thank you

**Comment by Gobliins** [comment]
[psoutput.txt](https://github.com/profesorfalken/jProcesses/files/229826/psoutput.txt)

You can remove the file after you downloaded it.

**Comment by profesorfalken** [comment]
Done, I will take it a look this evening (FR time)
Thank you

**Comment by profesorfalken** [comment]
Hi again.

I cannot reproduce it even forcing your output. 

I have added some logs (System.out) and a fix try into the snapshot version. 
Can you check it and tell me the results? If it still blocks it is interesting to know in what output line it gets stuck.
https://oss.sonatype.org/content/repositories/snapshots/org/jprocesses/jProcesses/1.4-SNAPSHOT/jProcesses-1.4-20160421.192204-2.jar

Thank you

**Comment by profesorfalken** [comment]
Have you take a look? I have created the 1.4 version that tries to fix this. 

Tell me if it does not work and I will make a fix version.

**Comment by Gobliins** [comment]
Hi, i used the version from your url and strangely it worked...
Sorry for late reply, i had a very busy week.

**Comment by profesorfalken** [comment]
I am glad to hear that I worked. I suppose that it is thanks to the fix I have done in reading method, adding a buffer flush before starting reading process. 

This fix is included in v1.4 so it should be alright. 

Thanks for reporting the problem.

**Comment by Gobliins** [comment]
Hi i recently upgraded to 1.5 and the same Problem seems to appear under linux again.

This is the exception:
CException in thread "de.planta.server.threads.MultipleSessionKillThread" java.lang.ArrayIndexOutOfBoundsException: 1
        at org.jutils.jprocesses.info.UnixProcessesService.loadFullCommandData(UnixProcessesService.java:160)
        at org.jutils.jprocesses.info.UnixProcessesService.parseList(UnixProcessesService.java:69)
        at org.jutils.jprocesses.info.AbstractProcessesService.getList(AbstractProcessesService.java:56)
        at org.jutils.jprocesses.info.AbstractProcessesService.getList(AbstractProcessesService.java:43)
        at org.jutils.jprocesses.JProcesses.getProcessList(JProcesses.java:41)
        at de.planta.server.webservice.SessionKill.killSessionMulti(SessionKill.java:97)
        at de.planta.server.threads.MultipleSessionKillThread.processPending(MultipleSessionKillThread.java:75)
        at de.planta.server.threads.MultipleSessionKillThread.run(MultipleSessionKillThread.java:37)
        at java.lang.Thread.run(Thread.java:745)

**Comment by profesorfalken** [comment]
Hello, 

Ok, I will take a look. 
Meanwhile you can continue to use the 1.4.x version as a workaround.

Thanks

**Comment by Gobliins** [comment]
Ok i must pu this issue on wait, i set up another linux server and the tool is wokring.. it must be an issue of the environment. I will try to figure it out what the problem is.

**Comment by Gobliins** [comment]
It seems like there is a deadlock in:

private static String readData(Process process)

on 

process.waitFor()

**Comment by profesorfalken** [comment]
Hello, 

No, wait. In fact what blocks it is the standard console, which hangs when the buffer is not fully flushed (maybe because there is too much output, as you say).

In the 1.4.2 I had done a fix for this, but in 1.5 I had removed it since I thought that it was not necessary anymore, but I was mistaken. This evening I will generate an snapshot for you to test and if it work I will generate a 1.5.1 with the fix.

Thanks

**Comment by profesorfalken** [comment]
Voilà:

https://oss.sonatype.org/content/repositories/snapshots/org/jprocesses/jProcesses/1.6-SNAPSHOT/jProcesses-1.6-20160607.184046-2.jar

**Comment by Gobliins** [comment]
Nice one, i will test it http://static.fjcdn.com/pictures/Soon_15db89_2831712.jpg

**Comment by Gobliins** [comment]
Ok i tried the snapshot but unfortunetly i still get the same deadlock.

**Comment by profesorfalken** [comment]
Ok it is strange since I have apply exactly the same fix that the last time. Now I have try it in an different way and added log to know in what process it blocks. Could you test it?

https://oss.sonatype.org/content/repositories/snapshots/org/jprocesses/jProcesses/1.6-SNAPSHOT/jProcesses-1.6-20160608.163059-3.jar

**Comment by Gobliins** [comment]
Maybe i was lucky the last time and on the machine weren't that much processes...
Ok it seems like it's working under linux now with the new Snapshot, thx.

One thing, could you add the "lstart" parameter in the ps options in the UnixProcessService so i can retrieve the full date or provide me with some source code for this snapshot so i can do it myself?

BTW: The JProcesses.getProcessList(String name) seems to be also problematic under windows as well as under linux. Under Linux i got a IndexArrayOutOfBoundsException and under Windows with WMI4J 1.3 i got a NoSuchMethodException and with WMI4J 1.4.1 -> java.lang.NoClassDefFoundError: com/google/common/base/Joiner 

Shall i open an issue for this?

**Comment by profesorfalken** [comment]
No problem with the lstart. I had planned to do it maybe this weekend. 

Can you open a new issue for the NoClassDefFoundError? Are you building with maven?

**Comment by profesorfalken** [comment]
Can you check this version to see if the deadlock is solved?

https://oss.sonatype.org/content/repositories/snapshots/org/jprocesses/jProcesses/1.6-SNAPSHOT/jProcesses-1.6-20160610.200328-4.jar

**Comment by Gobliins** [comment]
Hello again, i was not in office for a week, i will test it today.

**Comment by Gobliins** [comment]
Deadlock still exists when Console output gets to big.

Interesting behaviour, when i hit the strg+c button, the concole output is displayed before the program is interrupted.

**Comment by Gobliins** [comment]
Hi, i have made a fix to the problem:

 public static String executeCommand(String... command) {
        String commandOutput = null;

        try {
            //Process process = Runtime.getRuntime().exec(command);
            
            ProcessBuilder processBuilder = new ProcessBuilder(command);
            processBuilder.redirectErrorStream(true); // redirect error stream to output stream
            
            commandOutput = readData(processBuilder.start());
        } catch (IOException ex) {
            Logger.getLogger(ProcessesUtils.class.getName()).log(Level.SEVERE, null, ex);
        }

        return commandOutput;
    }


...


private static String readData(Process process) {
        StringBuilder commandOutput = new StringBuilder();
        BufferedReader processOutput = null;

        try {
        	processOutput = new BufferedReader(new InputStreamReader(process.getInputStream()));
        	 
        	//Fix for possible hanging. Flush output before start reading
            process.getOutputStream().flush();

            String line;
            while ((line = processOutput.readLine()) != null) {
                if (!line.isEmpty()) {
                    commandOutput.append(line).append(CRLF);
                }
            }
            
//            if (process.exitValue() == 0) {
//                processOutput = new BufferedReader(new InputStreamReader(process.getInputStream()));
//            } else {
//                processOutput = new BufferedReader(new InputStreamReader(process.getErrorStream()));
//            }

            
        } catch (IOException ex) {
            Logger.getLogger(ProcessesUtils.class.getName()).log(Level.SEVERE, null, ex);
        } finally {
            try {
                if (processOutput != null) {
                    processOutput.close();
                }
            } catch (IOException ioe) {
                Logger.getLogger(ProcessesUtils.class.getName()).log(Level.SEVERE, null, ioe);
            } finally {
            	try {
                    process.waitFor();
                } catch (InterruptedException ex) {
                    Logger.getLogger(ProcessesUtils.class.getName()).log(Level.SEVERE, null, ex);
                }
            }
        }

**Comment by Gobliins** [comment]
Basically i have moved the:

process.waitFor() behind the reading, because to me it didnt make sense to wait at the previous point.
also i have redirected both streams into one output which is not nice but i think there was no difference in handling the error stream at all.

i would be glad if you adapt / also improve this code into your project.

**Comment by profesorfalken** [comment]
Hello, 

I am glad you have find a solution. I am very busy these weeks but I will find a moment to fix it based in your comments.

Thanks for your help!

**Comment by profesorfalken** [comment]
OK for your solution using the ProcessBuilder. 
I will integrate it in a future 1.6 this weekend

Thanks again

**Comment by profesorfalken** [comment]
I have already published the 1.6 version so I close this issue.
